Central-Server/API/internal/threads.py

```diff
@@ -0,0 +1,398 @@
+            '/': {'content_type': 'text/html', 'filename': './static/ws_page.html'}
+        })
+
+        print("(DS-socketio) initializing server..", flush=True)
+
+    def run(self):
+        """This function is called to initialize the thread."""
+        print("(DS-socketio) DS Websocket initialized on " + str(self.websocket_port), flush=True)
+        eventlet.wsgi.server(eventlet.listen(('', self.websocket_port)), self.socketio_app) # Start WSGI server for websocket
+
+
+class BoardThread(threading.Thread): 
+    """Signifies a single datastreamer board connection."""
+    packet_buffer = None # The output buffer of this thread
+
+    def __init__(self, thread_name: str, thread_id: str, board_id: int, communication_channel: communication_interface.CommunicationChannel, pop_back_callback: callable) -> None: 
+        """Generates a datastreamer board thread
+        @thread_name: The name to give to this thread (arbitrary)
+        @thread_id: The ID for this thread (must be unique!)
+        @board_id: The ID of the board to which this thread is connected (must be unique!)
+        @communication_channel: The method that is used to communicate with the Datastreamer
+        @pop_back_callback: The function to be called when this thread needs to pop a job back in the queue."""
+        threading.Thread.__init__(self) 
+        self.thread_name = thread_name 
+        self.communication_channel = communication_channel
+        self.thread_id = thread_id 
+        self.cur_job_config: packets.DataPacket = None
+        self.has_job_config = False
+        self.running = True # 
+        self.packet_buffer = packets.DataPacketBuffer()
+        self.board_id = board_id
+        self.is_ready = False # True when in READY state (? TODO: check if this is true)
+        self.job_running = False # True when actively running job
+        self.board_type = ""
+        self.last_check = time.time()
+        self.callback = pop_back_callback
+        self.job_status: packets.JobStatus = None
+
+ 
+    def can_take_job(self):
+        """Whether the board associated to this thread is ready to take a job"""
+        return not self.has_job_config and self.is_ready
+
+    def take_job(self, config: packets.DataPacket):
+        """Assign the given job config to this board
+        @config: The JOB packet that contains this config"""
+        # Set all flags
+        self.cur_job_config = config
+        self.has_job_config = True
+
+        # Update the job status in the database
+        conn = database.connect()
+        cursor = conn.cursor()
+        cursor.execute("UPDATE hilsim_runs set run_status = %s, run_start = now() where run_id = %s",
+                       (jobs.JobStatus.RUNNING.value, config.data["job_data"]["job_id"]))
+        conn.commit()
+
+    def terminate(self) -> packets.DataPacket:
+        """Drop this thread"""
+        # TODO push current job to front of queue

```

Comment:
remove this misleading comment, we have already done this

Reply:
It doesn't seem to always work, I'll change it to say that it needs more testing.